### PR TITLE
Domain Picker: Add integration tests

### PR DESCRIPTION
**Pull request by StefanNieuwenhuis**
### Testing approach proposal

1. Write & Finalize our test scenarios.
2. Generate mock data.
3. Write the tests according to the AAA (Arrange-Act-Assert) pattern.
4. Run, Rinse & Repeat

#### Write & Finalize our test scenarios
In #50888 Acceptance Criteria, we have listed the testing scenarios that generate enough confidence in our code. Before we start coding, we have to agree and finalize them.

#### Generate mock data
Before we start writing tests, we have to mock API calls to not depend on API availability and state for passing/failing tests and possibly mock browser APIs (for example, local storage) since they simply do not exist in `jest`. 

But what to mock? There are three potential categories to be mocked:
1. Hooks.
2. Context Providers.
3. Component's Properties.

All managed dependencies (i.e. out-of-process dependencies we have full control over) **should not be mocked** but used as-is. Unmanaged dependencies should be mocked. Managed dependencies sometimes exhibit properties of both managed and unmanaged dependencies. In that case, we can replace the observable parts (i.e. exposed public API) with mocks. 

In this case, we consider hooks, context providers & component's props as unmanaged dependencies and therefore mock them.

#### Write the tests according to the AAA pattern
I think it's best to follow the same principles as we did in #50492

1. use 2 layers of `describe()`:
    1. The outer `describe()` declares the integration test (which flow is being tested).
    2. The inner `describe()` declares which part of the flow is being tested.
2. We use `it(...)` for all test cases.
3. All test cases should start with `should` as described in [BDD (language) concepts](https://softwareengineering.stackexchange.com/a/152233)

### Testing instructions

*

Related to #50888

### CODE REVIEW

**Comment by StefanNieuwenhuis** [comment]
1. Mock all the hooks in https://github.com/Automattic/wp-calypso/tree/fix/hide-ctas-from-success/packages/domain-picker/src/hooks
2. Mock the context (any context providers necessary)
3. Pass mocks as properties.

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	useDomainSuggestions: () => mockUseDomainSuggestionsResult,
+} ) );
+
+beforeEach( () => {
+	jest.clearAllMocks();
+} );
+
+describe( 'Search for a domain and select a suggestion', () => {
+	it( 'should search for a domain', () => {
+		const { getByText } = render(
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx]: According to [Kent C. Dodds](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen), we have to use `screen`.

File: packages/domain-picker/test/use-your-domain-item.tsx [hunk 2]
```diff
@@ -7,29 +7,28 @@
...
+	it( 'should fire onClick callback when clicked', () => {
 		const onClick = jest.fn();
 
 		render( <UseYourDomainItem onClick={ onClick } /> );
 
 		fireEvent.click( screen.getByText( 'Use a domain I own' ) );
-		expect( onClick ).toHaveBeenCalled();
+
+		// TODO: why is it called twice?
+		expect( onClick ).toHaveBeenCalledTimes( 2 );
```
Comment (ciampo) [inline packages/domain-picker/test/use-your-domain-item.tsx]: Change back to 1 once https://github.com/Automattic/wp-calypso/pull/51074 is merged

**Comment by ciampo** [comment]
Rebased and updated after #51074 was merged

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import { DataStatus, DomainSuggestion } from '@automattic/data-stores/src/domain-suggestions';
+
+/**
+ * Internal dependencies
+ */
+import '../matchMedia.mock';
+import DomainPicker from '../src/domain-picker';
+import { useDomainSuggestions } from '../src/hooks/use-domain-suggestions';
+
+// TODO: should we extract to separate mock data file/folder?
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx:15]: I think so. We could use a `__mocks__` folder to mock the different hooks/modules (https://jestjs.io/docs/manual-mocks).

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+			mockUseDomainSuggestionsResult.allDomainSuggestions.length
+		);
+
+		// Check if callback is called correctly
+		fireEvent.click( selectDomainButtons[ 0 ] );
+
+		expect( onDomainSelectCallback ).toHaveBeenCalledTimes( 1 );
+		expect( onDomainSelectCallback ).toHaveBeenCalledWith(
+			mockUseDomainSuggestionsResult.allDomainSuggestions[ 0 ]
+		);
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx:79]: We can replace these assertions with Jest's `mock` object.

```suggestion
	expect(onDomainSelectCallback.mock.calls).toEqual(mockUseDomainSuggestionsResult.allDomainSuggestions[ 0 ]);
```

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+				analyticsFlowId="12345"
+				initialDomainSearch="water"
+				itemType="button"
+			/>
+		);
+
+		const selectDomainButtons = getAllByText( 'Select', { exact: true, selector: 'button' } );
+
+		// Expect as many buttons as there are suggestion items in the data
+		expect( selectDomainButtons ).toHaveLength(
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx:69]: I think we're testing implementation details here, hence this assertion can be removed. WDYT?

File: packages/data-stores/src/domain-suggestions/index.ts [hunk 6]
```diff
@@ -15,6 +15,7 @@ import type { DispatchFromMap, SelectFromMap } from '../mapped-types';
 import { controls } from '../wpcom-request-controls';
 
 export * from './types';
+export * from './constants';
```
Comment (StefanNieuwenhuis) [inline packages/data-stores/src/domain-suggestions/index.ts:18]: Scope creep?

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ * External dependencies
+ */
+import * as React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+import { DataStatus, DomainSuggestion } from '@automattic/data-stores/src/domain-suggestions';
+
+/**
+ * Internal dependencies
+ */
+import '../matchMedia.mock';
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx:11]: Why do we need to mock `window.matchMedia`?

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	errorMessage: null,
+	state: DataStatus.Success,
+	retryRequest: jest.fn(),
+};
+
+jest.mock( '../src/hooks/use-domain-suggestions', () => ( {
+	useDomainSuggestions: () => mockUseDomainSuggestionsResult,
+} ) );
+
+beforeEach( () => {
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/search-and-select-domain-suggestion.tsx]: Just for structure, I would opt to move this inside the `before`. WDYT?

File: packages/domain-picker/test/use-your-domain-item.tsx [hunk 9]
```diff
@@ -7,29 +7,37 @@
...
 
-	test( 'Component fires onClick callback onClick', () => {
+	it( 'should fire onClick callback when the container is clicked', () => {
 		const onClick = jest.fn();
 
 		render( <UseYourDomainItem onClick={ onClick } /> );
 
-		fireEvent.click( screen.getByText( 'Use a domain I own' ) );
-		expect( onClick ).toHaveBeenCalled();
+		fireEvent.click( screen.getByTestId( 'use-domain-i-own-wrapper' ) );
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/test/use-your-domain-item.tsx]: According to [Kent C. Dodds](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-the-wrong-query), a better alternative for `getByTestId` is `getByRole`. This query resembles the user's behavior much better and that's what we have to aim for in integration testing. 

We can query for the `ArrowButton` inside the `UseYourDomainItem` since that event bubbles to the container (see #51074). See https://testing-library.com/docs/queries/about/#priority for more information.

**Review by StefanNieuwenhuis — COMMENTED** [review]
I left some comments :)

File: packages/data-stores/src/domain-suggestions/index.ts (hunk 6 above)
Comment (ciampo) [reply packages/data-stores/src/domain-suggestions/index.ts:18]: It's necessary for importing this line https://github.com/Automattic/wp-calypso/pull/50889/files#diff-81e5b2095be6643ef81b316eb251c80e5e29b24af18bbae45c019d70edd95ebaR6

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 7 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx:11]: Because of the `useViewportMatch` hook used by the domain suggestion item — without mocking `window.matchMedia`, the test would fail

https://github.com/Automattic/wp-calypso/blob/4edfbd6e9347ea3beeb77f5e1273caebc2c3d2bf/packages/domain-picker/src/domain-picker/suggestion-item.tsx#L72

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 8 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx]: Assuming you meant `describe` instead of `before`, I did that in https://github.com/Automattic/wp-calypso/pull/50889/commits/155365f4c162572170429609a2dbcdff7b952289

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 1 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx]: Done in https://github.com/Automattic/wp-calypso/pull/50889/commits/6d2c150fcd4d07edfab67e2892560aa37947a851

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 5 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx:69]: I thought this would be a good assertion — we need to know that the user is seeing and is able to pick any of the possible domain suggestions that the APIs returned.

This should potentially even become a separate test, since I believe that we show on screen only the first 5 suggestions (and a "see more" button).

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 4 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx:79]: I think that the current code is much easier to read (and will output a much readable error when it fails).

On top of that, your suggestion doesn't ensure that `onDomainSelectCallback` gets called only once

File: packages/domain-picker/test/use-your-domain-item.tsx (hunk 9 above)
Comment (ciampo) [reply packages/domain-picker/test/use-your-domain-item.tsx]: We can't `getByRole` in this case, since the underlying element is a simple `div` without any role. I agree that the markup needs improving, but that should be treated as a separate issue.

In the meantime, I still went ahead and tried to use better alternatives in the tests:

- I was able to use `getByRole` when selecting buttons (https://github.com/Automattic/wp-calypso/pull/50889/commits/261711b74db3ce4d8a9cfb020393afa4fbc992ad)
- For the specific case highlighted by this comment, I switched from `getByTestId` to `getByText` — at least a way of testing that is closer to how the user interacts with the component (https://github.com/Automattic/wp-calypso/pull/50889/commits/617c7dc0bf34ad54de44c2c377ebc51cc5a5bf58)

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 5 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx:69]: As agreed, this check will be removed from this specific test and will be moved to a unit test in a future issue/PR. For now, we're going to add a TODO comment

File: packages/domain-picker/test/search-and-select-domain-suggestion.tsx (hunk 3 above)
Comment (ciampo) [reply packages/domain-picker/test/search-and-select-domain-suggestion.tsx:15]: Agreed — this will be worked on in a separate issue/PR

**Review by ciampo — APPROVED** [review]
🚢
